File: ios/Sources/OptimoveSDKPlugin/OptimoveSDK.swift
import Foundation
import OptimoveSDK

/// Callback type for delivering events from native to the plugin bridge
public typealias OptimoveEventCallback = (_ eventName: String, _ data: [String: Any]) -> Void

@objc public class OptimoveSDKImpl: NSObject {

    private static var pendingPush: PushNotification?
    private static var pendingDdl: DeepLinkResolution?

    private static let sdkVersion = "1.0.0"
    private static let sdkTypeCapacitor = 107
    private static let runtimeTypeCapacitor = 4

    private var eventCallback: OptimoveEventCallback?

    public func setEventCallback(_ callback: @escaping OptimoveEventCallback) {
        self.eventCallback = callback
    }

    // MARK: - Initialization

    public func initialize(
        optimoveCredentials: String?,
        optimoveMobileCredentials: String?,
        inAppConsentStrategy: String,
        enableDeferredDeepLinking: Bool,
        ddlCname: String?,
        enableDelayedInitialization: Bool,
        delayedRegion: String?,
        delayedEnableOptimove: Bool,
        delayedEnableOptimobile: Bool
    ) {
        var builder: OptimoveConfigBuilder

        if enableDelayedInitialization {
            let regionStr = delayedRegion ?? "EU"
            let region: OptimobileConfig.Region
            switch regionStr {
            case "DEV": region = .DEV
            case "US": region = .US
            default: region = .EU
            }

            var featureSet: Feature = []
            if delayedEnableOptimove { featureSet.insert(.optimove) }
            if delayedEnableOptimobile { featureSet.insert(.optimobile) }

            builder = OptimoveConfigBuilder(region: region, features: featureSet)
        } else {
            guard optimoveCredentials != nil || optimoveMobileCredentials != nil else {
                print("OptimoveSDK: No credentials provided in capacitor.config")
                return
            }
            builder = OptimoveConfigBuilder(
                optimoveCredentials: optimoveCredentials,
                optimobileCredentials: optimoveMobileCredentials
            )
        }

        // Standard init without optimobile
        if optimoveMobileCredentials == nil && !enableDelayedInitialization {
            Optimove.initialize(with: builder.build())
            return
        }

        // Delayed init with optimobile disabled
        if enableDelayedInitialization && !delayedEnableOptimobile {
            Optimove.initialize(with: builder.build())
            return
        }

        // Push handlers
        builder.setPushOpenedHandler { [weak self] notification in
            guard let self = self, let cb = self.eventCallback else {
                OptimoveSDKImpl.pendingPush = notification
                return
            }
            cb("pushOpened", OptimoveSDKImpl.pushNotificationToDict(notification))
        }

        if #available(iOS 10, *) {
            builder.setPushReceivedInForegroundHandler { [weak self] notification, completionHandler in
                self?.eventCallback?("pushReceived", OptimoveSDKImpl.pushNotificationToDict(notification))
                completionHandler(.alert)
            }
        }

        // In-app consent
        switch inAppConsentStrategy {
        case "auto-enroll":
            builder.enableInAppMessaging(inAppConsentStrategy: .autoEnroll)
        case "explicit-by-user":
            builder.enableInAppMessaging(inAppConsentStrategy: .explicitByUser)
        default:
            break
        }

        // In-app deep link handler
        builder.setInAppDeepLinkHandler { [weak self] data in
            let dict: [String: Any] = [
                "deepLinkData": data.deepLinkData as Any,
                "messageId": data.messageId,
                "messageData": data.messageData as Any
            ]
            self?.eventCallback?("inAppDeepLink", dict)
        }

        // Deferred deep linking
        if enableDeferredDeepLinking {
            let ddlHandler: DeepLinkHandler = { [weak self] resolution in
                guard let self = self, let cb = self.eventCallback else {
                    OptimoveSDKImpl.pendingDdl = resolution
                    return
                }
                cb("deepLink", OptimoveSDKImpl.ddlResolutionToDict(resolution))
            }

            if let cname = ddlCname {
                builder.enableDeepLinking(cname: cname, ddlHandler)
            } else {
                builder.enableDeepLinking(ddlHandler)
            }
        }

        // Override install info for Capacitor
        overrideInstallInfo(builder: builder)

        Optimove.initialize(with: builder.build())

        OptimoveInApp.setOnInboxUpdated { [weak self] in
            self?.eventCallback?("inAppInboxUpdated", [:])
        }
    }

    // MARK: - Pending Events (Cold Start)

    public func deliverPendingEvents() {
        if let push = OptimoveSDKImpl.pendingPush {
            eventCallback?("pushOpened", OptimoveSDKImpl.pushNotificationToDict(push))
            OptimoveSDKImpl.pendingPush = nil
        }
        if let ddl = OptimoveSDKImpl.pendingDdl {
            eventCallback?("deepLink", OptimoveSDKImpl.ddlResolutionToDict(ddl))
            OptimoveSDKImpl.pendingDdl = nil
        }
    }

    // MARK: - User Identification

    public func setCredentials(optimoveCredentials: String?, optimobileCredentials: String?) {
        Optimove.setCredentials(
            optimoveCredentials: optimoveCredentials,
            optimobileCredentials: optimobileCredentials
        )
    }

    public func setUserId(_ userId: String) {
        Optimove.shared.setUserId(userId)
    }

    public func setUserEmail(_ email: String) {
        Optimove.shared.setUserEmail(email: email)
    }

    public func registerUser(userId: String, email: String) {
        Optimove.shared.registerUser(sdkId: userId, email: email)
    }

    public func getVisitorId() -> String {
        return Optimove.getVisitorID()
    }

    public func signOutUser() {
        Optimove.shared.signOutUser()
    }

    // MARK: - Analytics

    public func reportEvent(name: String, parameters: [String: Any]) {
        Optimove.shared.reportEvent(name: name, parameters: parameters)
    }

    public func reportScreenVisit(screenTitle: String, screenCategory: String?) {
        Optimove.shared.reportScreenVisit(screenTitle: screenTitle, screenCategory: screenCategory)
    }

    // MARK: - Push

    public func pushRequestDeviceToken() {
        Optimove.shared.pushRequestDeviceToken()
    }

    public func pushUnregister() {
        Optimove.shared.pushUnregister()
    }

    // MARK: - In-App Messaging

    public func inAppUpdateConsent(consented: Bool) {
        OptimoveInApp.updateConsent(forUser: consented)
    }

    public func inAppGetInboxItems() -> [[String: Any?]] {
        let inboxItems = OptimoveInApp.getInboxItems()
        let formatter = ISO8601DateFormatter()
        formatter.timeZone = TimeZone(secondsFromGMT: 0)

        return inboxItems.map { item in
            [
                "id": item.id,
                "title": item.title,
                "subtitle": item.subtitle,
                "availableFrom": item.availableFrom.map { formatter.string(from: $0) },
                "availableTo": item.availableTo.map { formatter.string(from: $0) },
                "dismissedAt": item.dismissedAt.map { formatter.string(from: $0) },
                "isRead": item.isRead(),
                "sentAt": formatter.string(from: item.sentAt),
                "imageUrl": item.getImageUrl()?.absoluteString,
                "data": item.data
            ]
        }
    }

    public func inAppGetInboxSummary(completion: @escaping (Int, Int) -> Void, failure: @escaping () -> Void) {
        OptimoveInApp.getInboxSummaryAsync { summary in
            guard let summary = summary else {
                failure()
                return
            }
            completion(summary.totalCount, summary.unreadCount)
        }
    }

    public func inAppMarkAsRead(messageId: Int) -> Bool {
        let inboxItems = OptimoveInApp.getInboxItems()
        guard let msg = inboxItems.first(where: { $0.id == Int64(messageId) }) else {
            return false
        }
        return OptimoveInApp.markAsRead(item: msg)
    }

    public func inAppMarkAllInboxItemsAsRead() -> Bool {
        return OptimoveInApp.markAllInboxItemsAsRead()
    }

    /// Returns: 0 = FAILED, 1 = EXPIRED, 2 = PRESENTED, -1 = not found
    public func inAppPresentInboxMessage(messageId: Int) -> Int {
        let inboxItems = OptimoveInApp.getInboxItems()
        guard let msg = inboxItems.first(where: { $0.id == Int64(messageId) }) else {
            return 0 // FAILED
        }

        let presentationResult = OptimoveInApp.presentInboxMessage(item: msg)
        switch presentationResult {
        case .PRESENTED: return 2
        case .EXPIRED: return 1
        default: return 0
        }
    }

    /// Returns true if deleted, false if not found
    public func inAppDeleteMessageFromInbox(messageId: Int) -> Bool {
        let inboxItems = OptimoveInApp.getInboxItems()
        guard let msg = inboxItems.first(where: { $0.id == Int64(messageId) }) else {
            return false
        }
        return OptimoveInApp.deleteMessageFromInbox(item: msg)
    }

    // MARK: - Private Helpers

    private func overrideInstallInfo(builder: OptimoveConfigBuilder) {
        let runtimeInfo: [String: AnyObject] = [
            "id": OptimoveSDKImpl.runtimeTypeCapacitor as AnyObject,
            "version": "8.0.0" as AnyObject,
        ]
        let sdkInfo: [String: AnyObject] = [
            "id": OptimoveSDKImpl.sdkTypeCapacitor as AnyObject,
            "version": OptimoveSDKImpl.sdkVersion as AnyObject,
        ]
        builder.setRuntimeInfo(runtimeInfo: runtimeInfo)
        builder.setSdkInfo(sdkInfo: sdkInfo)

        var isRelease = true
        #if DEBUG
            isRelease = false
        #endif
        builder.setTargetType(isRelease: isRelease)
    }

    private static func pushNotificationToDict(_ notification: PushNotification) -> [String: Any] {
        let aps: [AnyHashable: Any] = notification.aps
        var alert: [String: String] = [:]
        if let a = aps["alert"] as? [String: String] {
            alert = a
        }

        return [
            "id": notification.id,
            "title": alert["title"] as Any,
            "message": alert["body"] as Any,
            "data": notification.data as Any,
            "url": notification.url?.absoluteString as Any,
            "actionId": notification.actionIdentifier as Any
        ]
    }

    private static func ddlResolutionToDict(_ resolution: DeepLinkResolution) -> [String: Any] {
        var urlString: String
        var resolutionStr: String
        var content: [String: Any?]?
        var linkData: [AnyHashable: Any?]?

        switch resolution {
        case .lookupFailed(let dl):
            urlString = dl.absoluteString
            resolutionStr = "LOOKUP_FAILED"
        case .linkNotFound(let dl):
            urlString = dl.absoluteString
            resolutionStr = "LINK_NOT_FOUND"
        case .linkExpired(let dl):
            urlString = dl.absoluteString
            resolutionStr = "LINK_EXPIRED"
        case .linkLimitExceeded(let dl):
            urlString = dl.absoluteString
            resolutionStr = "LINK_LIMIT_EXCEEDED"
        case .linkMatched(let dl):
            urlString = dl.url.absoluteString
            resolutionStr = "LINK_MATCHED"
            content = [
                "title": dl.content.title,
                "description": dl.content.description
            ]
            linkData = dl.data
        }

        return [
            "resolution": resolutionStr,
            "url": urlString,
            "content": content as Any,
            "linkData": linkData as Any
        ]
    }
}

File: android/src/main/java/com/capacitor/optimove/sdk/OptimoveSDK.java
package com.capacitor.optimove.sdk;

import android.content.Context;

import androidx.annotation.Nullable;

import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

import com.optimove.android.Optimove;
import com.optimove.android.optimobile.InAppInboxItem;
import com.optimove.android.optimobile.InAppInboxSummary;
import com.optimove.android.optimobile.OptimoveInApp;

import com.getcapacitor.JSObject;
import com.getcapacitor.JSArray;

import org.json.JSONObject;

/**
 * Implementation class containing all Optimove SDK logic.
 * The plugin bridge (OptimoveSDKPlugin) delegates to this class.
 */
public class OptimoveSDK {

    // ─── User Identification ──────────────────────────────

    public void setCredentials(@Nullable String optimoveCredentials, @Nullable String optimobileCredentials) {
        if (optimoveCredentials != null && optimoveCredentials.isEmpty()) optimoveCredentials = null;
        if (optimobileCredentials != null && optimobileCredentials.isEmpty()) optimobileCredentials = null;
        Optimove.getInstance().setCredentials(optimoveCredentials, optimobileCredentials);
    }

    public void setUserId(String userId) {
        Optimove.getInstance().setUserId(userId);
    }

    public void setUserEmail(String email) {
        Optimove.getInstance().setUserEmail(email);
    }

    public void registerUser(String userId, String email) {
        Optimove.getInstance().registerUser(userId, email);
    }

    @Nullable
    public String getVisitorId() {
        return Optimove.getInstance().getVisitorId();
    }

    public void signOutUser() {
        Optimove.getInstance().signOutUser();
    }

    // ─── Analytics ────────────────────────────────────────

    public void reportEvent(String eventName, @Nullable JSObject params) throws Exception {
        if (params == null) {
            Optimove.getInstance().reportEvent(eventName);
        } else {
            Optimove.getInstance().reportEvent(eventName, JsonUtils.toMap(params));
        }
    }

    public void reportScreenVisit(String screenName, @Nullable String screenCategory) {
        if (screenCategory == null) {
            Optimove.getInstance().reportScreenVisit(screenName);
        } else {
            Optimove.getInstance().reportScreenVisit(screenName, screenCategory);
        }
    }

    // ─── Push ─────────────────────────────────────────────

    public void pushRequestDeviceToken() {
        Optimove.getInstance().pushRequestDeviceToken();
    }

    public void pushUnregister() {
        Optimove.getInstance().pushUnregister();
    }

    // ─── In-App Messaging ─────────────────────────────────

    public void inAppUpdateConsent(boolean consented) {
        OptimoveInApp.getInstance().updateConsentForUser(consented);
    }

    public JSObject inAppGetInboxItems() throws Exception {
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));

        List<InAppInboxItem> items = OptimoveInApp.getInstance().getInboxItems();
        JSArray results = new JSArray();

        for (InAppInboxItem item : items) {
            JSObject mapped = new JSObject();
            mapped.put("id", item.getId());
            mapped.put("title", item.getTitle());
            mapped.put("subtitle", item.getSubtitle());
            mapped.put("isRead", item.isRead());
            mapped.put("sentAt", formatter.format(item.getSentAt()));

            Date availableFrom = item.getAvailableFrom();
            Date availableTo = item.getAvailableTo();
            Date dismissedAt = item.getDismissedAt();
            JSONObject data = item.getData();
            URL imageUrl = item.getImageUrl();

            mapped.put("data", data == null ? JSObject.NULL : data);
            mapped.put("imageUrl", imageUrl == null ? JSObject.NULL : imageUrl.toString());
            mapped.put("availableFrom", availableFrom == null ? JSObject.NULL : formatter.format(availableFrom));
            mapped.put("availableTo", availableTo == null ? JSObject.NULL : formatter.format(availableTo));
            mapped.put("dismissedAt", dismissedAt == null ? JSObject.NULL : formatter.format(dismissedAt));

            results.put(mapped);
        }

        JSObject ret = new JSObject();
        ret.put("items", results);
        return ret;
    }

    public void inAppGetInboxSummary(InboxSummaryCallback callback) {
        OptimoveInApp.getInstance().getInboxSummaryAsync((InAppInboxSummary summary) -> {
            if (summary == null) {
                callback.onFailure();
                return;
            }
            callback.onSuccess(summary.getTotalCount(), summary.getUnreadCount());
        });
    }

    public boolean inAppMarkAsRead(int messageId) {
        InAppInboxItem item = getInboxItemById(messageId);
        if (item == null) return false;
        return OptimoveInApp.getInstance().markAsRead(item);
    }

    public boolean inAppMarkAllInboxItemsAsRead() {
        return OptimoveInApp.getInstance().markAllInboxItemsAsRead();
    }

    /**
     * @return presentation result ordinal (0=FAILED, 1=EXPIRED, 2=PRESENTED)
     */
    public int inAppPresentInboxMessage(int messageId) {
        InAppInboxItem item = getInboxItemById(messageId);
        if (item == null) return 0; // FAILED
        return OptimoveInApp.getInstance().presentInboxMessage(item).ordinal();
    }

    public boolean inAppDeleteMessageFromInbox(int messageId) {
        InAppInboxItem item = getInboxItemById(messageId);
        if (item == null) return false;
        return OptimoveInApp.getInstance().deleteMessageFromInbox(item);
    }

    // ─── Helpers ──────────────────────────────────────────

    @Nullable
    private InAppInboxItem getInboxItemById(int id) {
        List<InAppInboxItem> inboxItems = OptimoveInApp.getInstance().getInboxItems();
        for (InAppInboxItem item : inboxItems) {
            if (item.getId() == id) return item;
        }
        return null;
    }

    public interface InboxSummaryCallback {
        void onSuccess(int totalCount, int unreadCount);
        void onFailure();
    }
}
